waspc/src/Wasp/Lib.hs

```diff
@@ -43,75 +45,84 @@ compile ::
-
-analyzeWaspProject ::
+  compileWarnings <- maybeToList <$> warnIfDotEnvPresent waspDir
+  appSpecOrCompileErrors <- analyzeProject waspDir options
+  compileWarningsAndErrors <- case appSpecOrCompileErrors of
+    Left analyzerErrors -> return (compileWarnings, analyzerErrors)
+    Right appSpec -> generateCode appSpec outDir options
+  return $ (compileWarnings, []) <> compileWarningsAndErrors
+
+analyzeProject ::
   Path' Abs (Dir WaspProjectDir) ->
   CompileOptions ->
-  IO ([CompileWarning], Either (NonEmpty CompileError) AS.AppSpec)
-analyzeWaspProject waspDir options = do
-  maybeWaspFilePath <- findWaspFile waspDir
-  appSpecOrAnalyzerErrors <- case maybeWaspFilePath of
-    Nothing -> return $ Left $ fromList ["Couldn't find a single *.wasp file."]
-    Just waspFilePath -> do
-      waspFileContent <- readFile (SP.fromAbsFile waspFilePath)
-      case Analyzer.analyze waspFileContent of
-        Left analyzeError ->
-          return $
-            Left $
-              fromList
-                [ showCompilerErrorForTerminal
-                    (waspFilePath, waspFileContent)
-                    (getErrorMessageAndCtx analyzeError)
-                ]
-        Right decls -> do
-          externalCodeFiles <-
-            ExternalCode.readFiles (CompileOptions.externalCodeDirPath options)
-          maybeDotEnvServerFile <- findDotEnvServer waspDir
-          maybeDotEnvClientFile <- findDotEnvClient waspDir
-          maybeMigrationsDir <- findMigrationsDir waspDir
-          maybeUserDockerfileContents <- loadUserDockerfileContents waspDir
-          return $
-            Right
-              AS.AppSpec
-                { AS.decls = decls,
-                  AS.externalCodeFiles = externalCodeFiles,
-                  AS.externalCodeDirPath = CompileOptions.externalCodeDirPath options,
-                  AS.migrationsDir = maybeMigrationsDir,
-                  AS.dotEnvServerFile = maybeDotEnvServerFile,
-                  AS.dotEnvClientFile = maybeDotEnvClientFile,
-                  AS.isBuild = CompileOptions.isBuild options,
-                  AS.userDockerfileContents = maybeUserDockerfileContents
-                }
-  analyzerWarnings <- warnIfDotEnvPresent waspDir
-  return (analyzerWarnings, appSpecOrAnalyzerErrors)
+  IO (Either [CompileError] AS.AppSpec)
+analyzeProject waspDir options = runExceptT $ do
+  waspFilePath <- ExceptT $ left pure <$> findWaspFile waspDir
+  declarations <- ExceptT $ left pure <$> analyzeWaspFileContent waspFilePath
+  ExceptT $ constructAppSpec waspDir options declarations
+
+generateCode ::
+  AS.AppSpec ->
+  Path' Abs (Dir ProjectRootDir) ->
+  CompileOptions ->
+  IO ([CompileError], [CompileWarning])
+generateCode appSpec outDir options = do
+  (generatorWarnings, generatorErrors) <- Generator.writeWebAppCode appSpec outDir (sendMessage options)
+  let filteredWarnings = generatorWarningsFilter options generatorWarnings
+  return (map show filteredWarnings, map show generatorErrors)
```

Comment:
oh no `map` is hurting my eyes aaaaaaaa, nothing can help now but one nice cold `<$>`!

kidding, do `map` if you like it.

Reply:
Ha! That line is probably something leftover I did many months ago :D You can do Shayne Haskell-n00b cleanup if you'd like and wanna be cool like Martin haha I approve 👍🏻